packages/kbn-system-loader/src/plugin_spec.ts

```diff
@@ -0,0 +1,73 @@
+import {
+  PluginsType,
+  PluginName,
+  PluginConfigPath,
+  KibanaClassPluginStatic,
+  KibanaPlugin,
+} from './plugin';
+
+const noop = () => {};
+
+function isPromise(obj: any) {
+  return (
+    obj != null && typeof obj === 'object' && typeof obj.then === 'function'
+  );
+}
+
+export class PluginSpec<C, D extends PluginsType, E> {
```

Comment:
Since you seem cool with the request to rename to `system-loader`... How do you feel about some internal renames that move away from the `Plugin` name and use `System` instead?

It doesn't matter a ton for the internal names, but for names that are exposed via the API I think this would be more consistent:

```js
import {
  SystemLoader, // instead of PluginSystem
  System // instead of PluginSpec
} from '@kbn/system-loader'

const loader = new SystemLoader((name, metadata) => ({
  // ...
}))

loader.addSystems([
  new System('foo', {
    dependencies: ['bar'],
    implementation: require('../../plugins/foo'),
    metadata: {
      configPath: ['foo']
    }
  }),
  new System('bar', {
    dependencies: [],
    // instead of `plugin`
    implementation: require('../../plugins/bar'),
    metadata: {
      configPath: ['bar']
    },
  }),
])

loader.startSystems()
// ...
loader.stopSystems()
```